src/complexity.cc

```diff
@@ -73,31 +73,41 @@ std::string GetBigOString(BigO complexity) {
 //   - time          : Vector containing the times for the benchmark tests.
 //   - fitting_curve : lambda expression (e.g. [](int64_t n) {return n; };).
 
-// For a deeper explanation on the algorithm logic, look the README file at
-// http://github.com/ismaelJimenez/Minimal-Cpp-Least-Squared-Fit
+// This algorithm works like this:
+//
+// Let the execution times t_i for input sizes n_i be the function
+// t_i = f(n_i) = c*g(n_i), where c is a coefficient. The coefficient
+// c is compued by assuming that t_i and f(n_i) are the same function
+// with different coefficients t_i = c_1*g(n_i) and f(n_i) = c_2*g(n_i).
+// Since t_i and f(n_i) are the same function, they only differ in
+// proportionality: c_2 = c_1 * g(n_i) /  g(n_i)
+//
+// If, t_i and f(n_i) is not the same function, the Root Mean Square
+// (RMS) error will be large and dominated by n.
+// Example t_i = c_1*O(n^2) and  f(n_i) = c_2*O(n*log(n)):
+// ||e|| = sqrt(1/numSamples*sum((t_i - f(n_i))^2))
+// ||e|| = sqrt(1/numSamples*sum((c_1*n_i^2 - c_2*n_i*log(n_i))^2))
+// ||e||^2*numSamples = sum((c_1*n_i^2 - c_2*n_i*log(n_i)^2)
+// which is dominated by the difference (n_i^2 - n_i*log(n_i))^2
 
 LeastSq MinimalLeastSq(const std::vector<int64_t>& n,
                        const std::vector<double>& time,
                        BigOFunc* fitting_curve) {
   double sigma_gn = 0.0;
-  double sigma_gn_squared = 0.0;
   double sigma_time = 0.0;
-  double sigma_time_gn = 0.0;
 
   // Calculate least square fitting parameter
   for (size_t i = 0; i < n.size(); ++i) {
     double gn_i = fitting_curve(n[i]);
     sigma_gn += gn_i;
-    sigma_gn_squared += gn_i * gn_i;
     sigma_time += time[i];
-    sigma_time_gn += time[i] * gn_i;
   }
 
   LeastSq result;
   result.complexity = oLambda;
 
   // Calculate complexity.
-  result.coef = sigma_time_gn / sigma_gn_squared;
```

Comment:
FWIW i'm somewhat sure that this is the correct (other than fp errors) implementation of ["Least squares fitting"](https://en.wikipedia.org/wiki/Least_squares#Least_squares,_regression_analysis_and_statistics)
This function is named `MinimalLeastSq`, and has a square in the denominator; in the RHS of the diff, there is no more square..

Reply:
Strange. I re-performed the calculation again an landed in the same result that the coefficient is estimated by the proportion between sum of measurement and model. Good to see a source stating the result with squares. It is unfortunate to see that they omit the full derivation.

Having thought about it some more, I would actually claim that these methods are equivalent. The coefficient in the current case is computed as the proportion between the sum of the area of rectangles (ti*ni) and the sum of the area of squares (ni*ni). This is the same as computing the proportion between the sum of  lines (ti) and the sum of other lines (ni). This is not generally true of course, but for real positive values it holds. It is also possible to arrive at this conclusion by formulating the least-squre equation and solve for the coefficient. 

In any case, I am happy if we pursue this PR or update the reference to the calculation with a link to your Wikipedia-article or even use the old authors new repository? Also a test-case for the computation of the coefficient could be nice if it is interesting?

Reply:
The derivation of the formula is quite straightforward: take the sum of squares _S_ in the Wikipedia article, derive it with respect to _k_ and equate the derivative with 0, you get the estimated _k_.

Using a similar analysis it seems that the quantity that you are minimizing is the sum of terms like `c * t_i - c^2 * gn_i / 2`.  Look at this expression carefully: it can be negative for outliers: outliers will actually _reduce_ the error.  Put it another way: what you are minimizing is not even mathematically a norm.  Therefore your claim that your method is "equivalent" to least square has no mathematical basis.  (It is _if_ you ignore outliers, but that's not an acceptable thing to do: outliers will occur and they should not break the statistical analysis.)

I don't think that this PR should go in.  I would welcome a reference to the Wikipedia article, though.